src/core/router/transports/ssu/session.cc

```diff
@@ -620,56 +621,71 @@ void SSUSession::ProcessRelayRequest(
   SendRelayIntro(session.get(), from);
 }
 
-// TODO(anonimal): bytestream refactor
 void SSUSession::SendRelayRequest(
-    std::uint32_t introducer_tag,
-    const std::uint8_t* introducer_key) {
-  auto address = context.GetRouterInfo().GetSSUAddress();
-  if (!address) {
-    LOG(error)
-      << "SSUSession:" << GetFormattedSessionInfo()
-      << __func__ << ": SSU is not supported";
-    return;
-  }
-  std::array<std::uint8_t, 96 + SSUSize::BufferMargin> buf{
-      {}};  // TODO(unassigned): document size values
-  auto payload = buf.data() + SSUSize::HeaderMin;
-  core::OutputByteStream::Write<std::uint32_t>(payload, introducer_tag);
-  payload += 4;
-  *payload = 0;  // no address
-  payload++;
-  core::OutputByteStream::Write<std::uint16_t>(payload, 0, false);  // port = 0
-  payload += 2;
-  *payload = 0;  // challenge
-  payload++;
-  memcpy(payload, (const std::uint8_t *)address->key, 32);
-  payload += 32;
-  core::OutputByteStream::Write<std::uint32_t>(
-      payload, core::Rand<std::uint32_t>());  // nonce
-  std::array<std::uint8_t, SSUSize::IV> iv;
-  kovri::core::RandBytes(iv.data(), iv.size());
-  auto relay_request = SSUPayloadType::RelayRequest;
-  if (m_State == SessionState::Established) {
-    FillHeaderAndEncrypt(
-        relay_request,
-        buf.data(),
-        96,
-        m_SessionKey,
-        iv.data(),
-        m_MACKey);
-  } else {
-    FillHeaderAndEncrypt(
-        relay_request,
-        buf.data(),
-        96,
-        introducer_key,
-        iv.data(),
-        introducer_key);
-  }
+    const std::uint32_t introducer_tag,
+    const std::uint8_t* introducer_key)
+{
+  auto* const address = context.GetRouterInfo().GetSSUAddress();
+  if (!address)
+    {
+      LOG(error) << "SSUSession:" << GetFormattedSessionInfo() << __func__
+                 << ": SSU is not supported";
+      return;
+    }
+
+  // Create message  // TODO(anonimal): move to packet writer
+  // TODO(unassigned): size if we include Alice's IP (see SSU spec, unimplemented)
+  core::OutputByteStream message(
+      SSUSize::RelayRequestBuffer + SSUSize::BufferMargin);  // TODO(anonimal): review buffer margin
+
+  // TODO(unassigned): Endianness is not spec-defined, assuming BE
+
+  // Skip header (written later)
+  message.SkipBytes(SSUSize::HeaderMin);
+
+  // Intro tag
+  message.Write<std::uint32_t>(introducer_tag);
+
+  // Address, port, and challenge (see SSU spec)
+  message.SkipBytes(4);
```

Comment:
IIUC, in the Java spec, the ip-size byte is followed by that many bytes of the actual IP.
Does this need to be accounted for here, or am I missing something?

Reply:
>am I missing something?

Yes, look more carefully.

- `96 bytes (no Alice IP included)`
- `In the current implementation, the IP length is always 0 and the port is always 0`.
- `Challenge is unimplemented, challenge size is always zero`

Reply:
Knew I had to be missing something, thanks for explaining.